kasa/klapprotocol.py

```diff
@@ -0,0 +1,487 @@
+for encryption.  The last 4 bytes of the initialisation vector
+are used as a sequence number that increments every time the
+client calls encrypt and this sequence number is sent as a
+url parameter to the device along with the encrypted payload
+
+https://gist.github.com/chriswheeldon/3b17d974db3817613c69191c0480fe55
+https://github.com/python-kasa/python-kasa/pull/117
+
+"""
+
+import asyncio
+import datetime
+import hashlib
+import logging
+import secrets
+import time
+from pprint import pformat as pf
+from typing import Any, Dict, Optional, Tuple, Union
+
+import httpx
+from cryptography.hazmat.primitives import hashes, padding
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+
+from .credentials import Credentials
+from .exceptions import AuthenticationException, SmartDeviceException
+from .json import dumps as json_dumps
+from .json import loads as json_loads
+from .protocol import TPLinkProtocol
+
+_LOGGER = logging.getLogger(__name__)
+logging.getLogger("httpx").propagate = False
+
+
+class TPLinkKlap(TPLinkProtocol):
+    """Implementation of the KLAP encryption protocol.
+
+    KLAP is the name used in device discovery for TP-Link's new encryption
+    protocol, used by newer firmware versions.
+    """
+
+    DEFAULT_PORT = 80
+    DEFAULT_TIMEOUT = 5
+    DISCOVERY_QUERY = {"system": {"get_sysinfo": None}}
+    KASA_SETUP_EMAIL = "kasa@tp-link.net"
+    KASA_SETUP_PASSWORD = "kasaSetup"  # noqa: S105
+    SESSION_COOKIE_NAME = "TP_SESSIONID"
+
+    def __init__(
+        self,
+        host: str,
+        credentials: Optional[Credentials] = None,
+        *,
+        timeout: Optional[int] = None,
+    ) -> None:
+        super().__init__(host=host, port=self.DEFAULT_PORT)
+
+        self.credentials = (
+            credentials
+            if credentials and credentials.username and credentials.password
+            else Credentials(username="", password="")
+        )
+
+        self._local_seed: Optional[bytes] = None
+        self.local_auth_hash = self.generate_auth_hash(self.credentials)
+        self.local_auth_owner = self.generate_owner_hash(self.credentials).hex()
+        self.kasa_setup_auth_hash = None
+        self.blank_auth_hash = None
+        self.handshake_lock = asyncio.Lock()
+        self.query_lock = asyncio.Lock()
+        self.handshake_done = False
+
+        self.encryption_session: Optional[KlapEncryptionSession] = None
+        self.session_expire_at: Optional[float] = None
+
+        self.timeout = timeout if timeout else self.DEFAULT_TIMEOUT
+        self.session_cookie = None
+        self.http_client: Optional[httpx.AsyncClient] = None
+
+        _LOGGER.debug("Created KLAP object for %s", self.host)
+
+    @staticmethod
+    def _sha256(payload: bytes) -> bytes:
+        return hashlib.sha256(payload).digest()
+
+    @staticmethod
+    def _md5(payload: bytes) -> bytes:
+        digest = hashes.Hash(hashes.MD5())  # noqa: S303
+        digest.update(payload)
```

Comment:
Please move these outside from the class to make them just regular functions, no need for them to be inside the class.

Reply:
Done